codewhisperer: create webview panel for security issue

ctlai95 opened the pull request:
## Problem

Webview panel is needed to show security issue details.

## Solution

Creates a webview panel class which:
- will open a tab beside the active editor
- will only have one tab open at a time
- is able to dynamically update its content with a new issue
- caches the issue so that serialization can load it again if vscode restarts

For the content:
- Uses markdown parser to render diff code block in html
- Uses `webview-ui-toolkit` design library (this requires a separate webpack config)

![image](https://github.com/aws/aws-toolkit-vscode/assets/16008563/cdae8c41-166b-4dd2-b9ea-eda4c2c98ee0)



## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

justinmk3 commented:
Using `webview-ui-toolkit` is a good idea, that helps to match vscode's theme.

Can the hover panel https://github.com/aws/aws-toolkit-vscode/pull/3884 provide an "Apply Fix" button? This webview seems redundant with the hover provider.

ctlai95 commented:
> Can the hover panel #3884 provide an "Apply Fix" button? This webview seems redundant with the hover provider.

@justinmk3 Yes, the hover panel also has an "Apply Fix" button that will link to the same command. But not all issues will have a fix so sometimes it will be hidden. As for redundancy, we wanted to keep the hover content as short as possible and link out to the webview for all the details.

![image](https://github.com/aws/aws-toolkit-vscode/assets/16008563/9c23856e-4d5d-48f1-840f-74d19fc6f296)

justinmk3 reviewed (dismissed):
Did you consider using a [CustomTextEditorProvider](https://code.visualstudio.com/api/extension-guides/custom-editors)? Or just a plain markdown file? That would avoid the new dependencies "markdown-it" and "highlight.js". 

Each webview is essentially a custom web application, with its own dependencies and behavior. That is a maintenance cost, and it's a const for users whom must learn the custom behavior of each of these custom web apps. For example, because this custom webview isn't a vscode text editor, none of the user's normal text editor shortcuts will work.

ctlai95 commented on package.json:
```diff
@@ -3572,7 +3573,7 @@
         "buildBrowser": "npm run clean && npm run buildScripts && webpack --config webpack.browser.config.js --mode production && npm run copyFiles",
         "compile": "npm run clean && npm run buildScripts && webpack --mode development && npm run copyFiles",
-        "watch": "npm run clean && npm run buildScripts && tsc -watch -p ./",
+        "watch": "npm run clean && npm run buildScripts && webpack --config webpack.webview.config.js --mode development && tsc -watch -p ./",
```
It's not needed for other webpack configs because those files are being emitted from the `tsc` command. And we can't rely on `tsc` for `webview.ts` because we want webpack to include its dependencies in the output

justinmk3 replied:
How are other developers expected to understand these details? The differences are not obvious, and this configuration is fragile (easily broken in the future).

Why can't all the webviews work the same?

ctlai95 replied:
Good point, let me see if there's another way to do this.

ctlai95 commented:
@justinmk3 Could you elaborate on how `CustomTextEditorProvider` might help in this case? If I understand it correctly, it allows creating webviews instead of normal text editor documents, but doesn't this require a separate file in the current workspace?

justinmk3 commented:
> As for redundancy, we wanted to keep the hover content as short as possible and link out to the webview for all the details.

Is that a strong reason to add a custom webapp and introduce 2 new dependencies? What is the UX if the hover contents are long? Is opening a new tab (after an extra click) a better UX?

> I think using plain markdown won't work as well, since we need some HTML elements like buttons for applying a fix and also to apply custom styling for certain elements.

vscode extensions can provde "codelenses" but I don't know if vscode's markdown Preview is extensible.

> Could you elaborate on how `CustomTextEditorProvider` might help in this case? If I understand it correctly, it allows creating webviews instead of normal text editor documents, but

> doesn't this require a separate file in the current workspace?

Where are you seeing that requirement ([docs](https://code.visualstudio.com/api/extension-guides/custom-editors))? vscode provides a [full example](https://github.com/microsoft/vscode-extension-samples/blob/a8a74c4d7773470e3d53cb89502b442579d5de84/custom-editor-sample/src/catScratchEditor.ts#L16).

But I'm not sure if CustomTextEditorProvider will fit your exact use-case.

ctlai95 commented on src/codewhisperer/models/model.ts:
```diff
@@ -196,7 +196,12 @@ export interface CodeScanIssue {
+
+export interface Remediation {
+    recommendation: { text: string; url: string }
+    suggestedFixes?: { description: string; code: string }[]
```
updated it

ctlai95 commented:
@justinmk3 @mj632 I removed webview-ui-toolkit dependency for now since I was using it mainly for the button component but we can achieve the same thing with pure css. Not worth having the extra webpack config imo.

I also refactored the class to use VueWebview and moved the front end part to vue. This should help conform to the rest of the extension's architecture as well as avoid introducing additional webview logic for this change. I still think we need to have `markdown-it` and `highlight.js` dependencies to render the code diff. Let me know what you guys think

ctlai95 commented on src/codewhisperer/views/securityIssue/vue/root.vue:
```diff
@@ -0,0 +1,202 @@
+})
+</script>
+
+<style>
```
`securityIssue.css` includes styles from vscode's markdown preview. The ones above are our custom ones, so I wanted to keep them separate

ctlai95 commented on src/codewhisperer/service/securityIssueHoverProvider.ts:
```diff
@@ -59,12 +59,14 @@ export class SecurityIssueHoverProvider implements vscode.HoverProvider {
+            `command:aws.codeWhisperer.openSecurityIssuePanel?${encodeURIComponent(JSON.stringify(issue))}`
+        )
+        const applyFixCommand = vscode.Uri.parse('command:aws.codeWhisperer.applySecurityFix')
+        markdownString.appendMarkdown(`[$(eye) View Details](${viewDetailsCommand} "Open security issue")\n`)
```
Sure I can do that
